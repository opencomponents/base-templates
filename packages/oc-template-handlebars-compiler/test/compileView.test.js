jasmine.DEFAULT_TIMEOUT_INTERVAL = 50000;

const _ = require('lodash');
const compileView = require('../lib/compileView.js');
const fs = require('fs-extra');
const path = require('path');

const componentPath = path.join(
  __dirname,
  '../../../mocks/handlebars-component'
);
const publishPath = path.join(componentPath, '_compile-view-package');
const publishFileName = 'template.js';

const options = {
  componentPackage: {
    oc: {
      files: {
        template: {
          src: 'template.hbs',
          type: 'handlebars'
        }
      }
    }
  },
  componentPath,
  publishPath,
  publishFileName
};

test('Should correctly compile the view', done => {
  compileView(options, (err, compiledViewInfo) => {
    const viewHashKey = compiledViewInfo.template.hashKey;
    compiledViewInfo.template.hashKey = 'dummyData';
    expect(compiledViewInfo).toMatchSnapshot();
    expect(
      fs
        .readFileSync(path.join(publishPath, publishFileName), 'UTF8')
        .replace(viewHashKey, 'dummyData')
    ).toMatchSnapshot();
    fs.removeSync(path.resolve(publishPath));
    done();
  });
});

test('When view not found should return error', done => {
  const options2 = _.cloneDeep(options);
  options2.componentPackage.oc.files.template.src = 'not-found.hbs';
  compileView(options2, (err, compiledViewInfo) => {
    expect(err).toMatchSnapshot();
    done();
  });
});

test('When view not valid should return error', done => {
  const spy = jest
    .spyOn(fs, 'readFile')
    .mockImplementation(
      jest.fn((a, b, cb) => cb(null, '<h1>{{something</h1>'))
    );

  compileView(options, (err, compiledViewInfo) => {
    expect(err).toMatchSnapshot();
    spy.mockRestore();
    done();
  });
});

test('When compiled view writing fails should return error', done => {
  const spy = jest
    .spyOn(fs, 'ensureDir')
    .mockImplementation(jest.fn((a, cb) => cb('sorry I failed')));

  compileView(options, (err, compiledViewInfo) => {
    expect(err).toMatchSnapshot();
    spy.mockRestore();
    done();
  });
});
