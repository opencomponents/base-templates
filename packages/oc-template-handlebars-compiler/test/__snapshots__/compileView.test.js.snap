// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly compile the view 1`] = `
Object {
  "template": Object {
    "hashKey": "dummyData",
    "src": "template.js",
    "type": "handlebars",
  },
}
`;

exports[`Should correctly compile the view 2`] = `"var oc=oc||{};oc.components=oc.components||{},oc.components[\\"dummyData\\"]={compiler:[8,\\">= 4.3.0\\"],main:function(c,o,e,n,a){return\\"<h1>Hello handlebars!</h1>\\"},useData:!0};"`;

exports[`When compiled view writing fails should return error 1`] = `"template.hbs compilation failed - sorry I failed"`;

exports[`When view not found should return error 1`] = `"file not-found.hbs not found"`;

exports[`When view not valid should return error 1`] = `
"template.hbs compilation failed - Error: Parse error on line 1:
<h1>{{something</h1>
------^
Expecting 'ID', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', got 'INVALID'"
`;
